File: Algo/Leetcode/946ValidStackSeq.cpp
class Solution {
public:
    bool validateStackSequences(vector<int>& pushed, vector<int>& popped) {
        stack<int> s;
        if(pushed.size()!=popped.size())
            return false;
        if(pushed.size() == 0)
            return true;

        int i=0; int j=0;
        while(i < pushed.size()){
            s.push(pushed[i++]);
            while(!s.empty() && s.top()==popped[j]){
                s.pop();
                j++;
            }
        }
        return s.empty();
    }
};

File: Algo/Leetcode/946ValidStackSeq.py
class Solution(object):
    def validateStackSequences(self, pushed, popped):
        """
        :type pushed: List[int]
        :type popped: List[int]
        :rtype: bool
        """
        stack = []
        n = len(pushed)
        i = 0
        j = 0
        while i < n+1 and j < n:
            if len(stack) > 0 and stack[-1] == popped[j]:
                stack.pop()
                j += 1
            elif i < n:
                stack.append(pushed[i])
                i += 1
            else:
                break

        return j == n